### DISCUSSION TITLE
Relation to nvim-lsp-installer

### DISCUSSION THREAD

**Discussion by JSchrtke** [mason-org/mason.nvim#20] [Q&A] [answered]
Hi there, just saw this recently. I was wondering how this relates to your older 'nvim-lsp-installer'? Is this meant to replace that eventually? Since this can also install lsp servers and such...

**Comment by williamboman** [comment, marked as answer]
Hello! I opened this repo up for early access and potentially get some feedback on usage, wasn't expecting much attention at first. I'll clarify the idea and rationale behind this plugin (and how it relates to nvim-lsp-installer) soon!

**Reply by JSchrtke** [reply]
Alright, cool. I'll definitely do some testing, especially since I'm on windows, and things usually have some issues there.

**Reply by williamboman** [reply]
Great! I need to document how the linking works on Windows as it will currently not play nicely with some things. Short description: executables will be "linked" as `.cmd` wrapper scripts in mason's `bin/` directory. These wrapper scripts will in turn execute the actual executable installed in `packages/<package_name>/...`.

Some utilities, libuv in particular (which neovim's RPC client, among others, uses), have problems finding `.cmd` executables in `PATH`. As it currently stands, server configurations in `lspconfig` that assume an `.exe` executable exists is not compatible with mason.nvim - the workaround is to use something like `:h exepath()`, like so:

```lua
lspconfig.sumneko_lua.setup {
  cmd = { vim.fn.exepath("lua-language-server") }
}
```

I'm leaning towards patching these server configs in the [`mason-lspconfig`](https://github.com/williamboman/mason.nvim/tree/alpha/lua/mason-lspconfig) extension.

null-ls [handles this for you already](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/fbb1929b29beff82e0fc495670f00ef4b3bcbcd3/lua/null-ls/loop.lua#L147), and I'm currently trying to discover other plugins that may run into these issues.

**Reply by williamboman** [reply]
I'd also like to see if this approach of "linking" executables makes sense for Windows devs, or if it's too foreign/problematic. I've only done cursory testing on Windows and things seem to work perfectly fine 🤷‍♂️

**Reply by JSchrtke** [reply]
funny you mention that, I'm typing an issue as we speak :D

Tldr:   Downloading file "https://github.com/golang/vscode-go/releases/download/v0.34.1/go-0.34.1.vsix"...
  ...ker\start\mason.nvim/lua/mason/core/installer/linker.lua:44: Link target "C:\\Users\\js\\AppData\\Local\\nvim-data\\mason\\packages\\go-debug-adapter\\go-debug-adapter" does not exist.

I'll leave the issue with the full details, I can also have a look what the problem is during the some time during the weekend

Edit: nice issue template btw, I don't even need to think anymore

**Reply by JSchrtke** [reply]
> I'd also like to see if this approach of "linking" executables makes sense for Windows devs, or if it's too foreign/problematic. I've only done cursory testing on Windows and things seem to work perfectly fine 🤷‍♂️

Sorry, forgot to answer regarding this one. I don't really see an issue with using .cmd/.bat scripts to invoke things. As for lspconfig only expecting .exe files; I assume this is a windows-only issue? As in, can lspconfig handle things like shell scripts to start the server? If so, I think the fix should really be upstream, such that lspconfig can also handle starting a language server via scripts (.cmd/.bat/.ps2). Anyway, I think an issue for keeping track of this is perhaps helpful, rather than this discussion?

**Reply by williamboman** [reply]
> I assume this is a windows-only issue?

If my memory doesn't fail me - it's really only a Window + libuv issue. So any plugins or Neovim builtins that leverages `uv_spawn` and doesn't make any considerations with regards to `.cmd`/`.bat` executables will not be compatible with mason executables. This is very easily remedied by `vim.loop.spawn(vim.fn.exepath("<executable>"))`, as done by null-ls.

> As in, can lspconfig handle things like shell scripts to start the server?
>  If so, I think the fix should really be upstream, such that lspconfig can also handle starting a language server via scripts (.cmd/.bat/.ps2).

lspconfig actually does use the shell to spawn all Node.js/npm-based servers - like `cmd.exe /C <executable>` (this is akin to how Node.js solves it as well via their high-level [`process.spawn` API](https://nodejs.org/api/child_process.html#shell-requirements)). So there's definitely a possibility to do this but the server configs in lspconfig are best-effort and unsupported (according to their README), I doubt they'll be interested in considering this tbh and it's not something I'll chase. I believe they only aim to support the installation method they've provided for each server (which usually is via a system package manager). Also see https://github.com/neovim/neovim/pull/18631#discussion_r887935051

>  Anyway, I think an issue for keeping track of this is perhaps helpful, rather than this discussion?

Maybe that'd be too much admin work, I think I'll settle with "patching" the incompatible lspconfig configurations inside the `mason-lspconfig` extension - that's easily done.

**Reply by JSchrtke** [reply]
Alright, sounds like you've got a good solution already then. Feel free to ping me for windows issues, I'd be glad to help.
